pallets/mixer/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+
+		#[weight = 0]
+		pub fn deposit(origin, mixer_id: T::GroupId, data_points: Vec<Data>) -> dispatch::DispatchResult {
+			let sender = ensure_signed(origin)?;
+			ensure!(Self::initialised(), Error::<T>::NotInitialised);
+			// get mixer info, should always exist if module is initialised
+			let mut mixer_info = Self::get_mixer(mixer_id)?;
+			// ensure the sender has enough balance to cover deposit
+			let balance = T::Currency::free_balance(&sender);
+			// TODO: Multiplication by usize should be possible
+			// using this hack for now, though we should optimise with regular multiplication
+			// `data_points.len() * mixer_info.fixed_deposit_size`
+			let deposit: BalanceOf<T> = data_points.iter()
+				.map(|_| mixer_info.fixed_deposit_size)
+				.fold(Zero::zero(), |acc, elt| acc + elt);
+			ensure!(balance >= deposit, Error::<T>::InsufficientBalance);
+			// transfer the deposit to the module
+			T::Currency::transfer(&sender, &Self::account_id(), deposit, AllowDeath)?;
+			// add elements to the mixer group's merkle tree and save the leaves
+			T::Group::add_members(Self::account_id(), mixer_id.into(), data_points.clone())?;
+			for i in 0..data_points.len() {
+				mixer_info.leaves.push(data_points[i]);	
+			}
+			MixerGroups::<T>::insert(mixer_id, mixer_info);
+			
+			Ok(())
+		}
+
+		#[weight = 0]
+		pub fn withdraw(
+			origin,
+			mixer_id: T::GroupId,
+			cached_block: T::BlockNumber,
+			cached_root: Data,
+			leaf_com: Commitment,
+			path: Vec<(Commitment, Commitment)>,
+			r_com: Commitment,
+			nullifier: Data,
+			proof_bytes: Vec<u8>
+		) -> dispatch::DispatchResult {
+			let sender = ensure_signed(origin)?;
+			ensure!(Self::initialised(), Error::<T>::NotInitialised);
+			let mixer_info = MixerGroups::<T>::get(mixer_id);
+			// check if the nullifier has been used
+			// Returns `()` if the nullifier has not been used
+			// otherwsie returns `AlreadyUsedNullifier` error from merkle groups pallet
+			T::Group::has_used_nullifier(mixer_id.into(), nullifier)?;
+			// Verify the zero-knowledge proof of membership provided
+			// Returns `()` if verification is successful
+			// Otherwise returns `Err` for failed verification / bad proof from merkle groups pallet
+			T::Group::verify_zk_membership_proof(
+				mixer_id.into(),
+				cached_block,
+				cached_root,
+				leaf_com,
+				path,
+				r_com,
+				nullifier,
+				proof_bytes,
+			)?;
+			// transfer the fixed deposit size to the sender
+			T::Currency::transfer(&Self::account_id(), &sender, mixer_info.fixed_deposit_size, AllowDeath)?;
+			// Add the nullifier on behalf of the module
+			T::Group::add_nullifier(Self::account_id(), mixer_id.into(), nullifier)
+		}
+
+		#[weight = 0]
+		pub fn initialize(origin) -> dispatch::DispatchResult {
+			ensure!(!Self::initialised(), Error::<T>::AlreadyInitialised);
+			let _ = ensure_signed(origin)?;
+			Initialised::set(true);
+			let one: BalanceOf<T> = One::one();
+			let depth: u8 = <T as Config>::MaxTreeDepth::get();
```

Comment:
For performance reason, maybe we should use height 20, in which case we should make sure new trees are created when the old ones are full (this would require further changes, but first lmk your thoughts)

Reply:
I think this is fine for now. I have an idea for an optimisation to nullifier storage that will likely reduce our storage issues entirely, but it'll again push storage of the nullifier set offchain.